luaui/Widgets/dbg_widget_restriction_gameside.lua

```diff
@@ -0,0 +1,24 @@
+local side = "Gameside" -- copy to local luaui and rename to test this userside
+
+function widget:GetInfo()
+  return {
+    name      = "Widget Restricton "..side,
+    desc      = "Tests for widget restrictions on "..side,
+    author    = "Beherith",
+    date      = "2023.07.19",
+    license   = "GNU GPL v2",
+    layer     = 0,
+    enabled   = false  --  loaded by default?
+  }
+end
+-- the test case will be the usage of GetProjectilesInRectangle
+local sp = Spring.Utilities.Gametype.IsSinglePlayer()
+
+local spGetProjectilesInRectangle  = SpringRestricted.GetProjectilesInRectangle
+SpringRestricted = nil -- for some reason, sprung knows that this is a good idea, 
```

Comment:
this is a bad idea :Þ

Reply:
Why was it done though?

Reply:
other widgets could access it. the bad idea here is not having a whitelist, since you need to nil it in all widgets that get access to the table (which in your version is all fromzip widgets) and do it before any vfs.include (unless you poison vfs.include too)

Reply:
Ok, I have some questions then:

1. VFS.Include/LoadFile:

So even a user widget would get access to Spring.* by VFS.Including something, and e.g. returning the Spring table from there, because VFS.Include gets Spring.* from System.lua?

2. How could, in principle "other widgets access it" if the SpringRestricted table is not nilled? 

3. Do you have a better approach to reigning in user widgets access, or is this approach a good starting direction?

Reply:
1) `VFS.Include` doesn't get `Spring.*` from `system.lua`, it gets it from here https://github.com/beyond-all-reason/Beyond-All-Reason/blob/432f6cbbde1d02b7d22c8c20f5628885d1d68610/luaui/barwidgets.lua#L539-L542 . The reason I mentioned `VFS.Include` is that if a whitelisted widget does something like this:
```lua
local restricted = SpringRestricted.Bla
VFS.Include("my_wupget_config.lua")
SpringRestricted = nil -- too late!
```
then a user can just replace `my_wupget_config.lua` with one that has `GG.pwned = SpringRestricted` inside unless you also fortify `VFS.Include` (which is doable but also needs effort to do well unless you go ham).

2) `handler = true` gives you access to widget handler, from which you can reach other widgets. It could be disabled but offers some useful functionality that would ideally be replaced if you don't want to have a whitelist. For widgets that contain `VFS.Include` also the above. There's also `debug.bla` which could probably be used to access things, but here I think even `= nil` won't help so you'd probably need to block that one despite it also being useful for modders.

3) I am unaware of any better approach and I think in principle it should be viable if done comprehensively, although the reason it was done in ZK is mostly so that you can't reach those functions by accident or via basic bitch modding so I wouldn't mind too much if the current implementation was breakable by sufficiently determined people because being 100% tinker-proof wasn't the goal. Part of this is because of game differences (baseline ZK interface is very powerful so tinkering ultimately doesn't matter that much) and part because if you're l33t enough of a h4x0r to break it you can also just compile the engine and get unlimited hax power.